### PR TITLE
types/hardware: Add Infiniband definitions.

### PR DESCRIPTION
**Pull request by wdpypere**
We use these at UGent for maintaining IB HCA's.

### CODE REVIEW

**Comment by jrha** [comment]
@hpcugentbot test this please

**Comment by ned21** [comment]
Looks OK to me but tests failed on panlint bug?

File: quattor/types/hardware.pan [hunk 1]
```diff
@@ -75,6 +75,36 @@ type structure_bmc = {
...
 
 @documentation{
+    PCI addresses
+}
+type structure_pci = {
+    "vendor" ? long
+    "device" ? long
+    "class"  ? long
+};
+@docmentation{
```
Comment (stdweird) [inline quattor/types/hardware.pan]: insert newline

File: quattor/types/hardware.pan [hunk 2]
```diff
@@ -75,6 +75,36 @@ type structure_bmc = {
...
+    IB hardware address
+}
+type ibhwaddr = string;
```
Comment (stdweird) [inline quattor/types/hardware.pan]: no validation?

File: quattor/types/hardware.pan [hunk 3]
```diff
@@ -75,6 +75,36 @@ type structure_bmc = {
...
+@documentation{
```
Comment (stdweird) [inline quattor/types/hardware.pan]: insert newline

File: quattor/types/hardware.pan [hunk 4]
```diff
@@ -75,6 +75,36 @@ type structure_bmc = {
...
+    "media" ? string
+    "name" ? string
+    "pxe" ? boolean
+    "boot" ? boolean
+    "ca_name" ? string
+    "portnum" ? long(1..)
+    "hwaddr" ? ibhwaddr
+    "active" ? boolean
+    "pci" ? structure_pci
+    "guid" ? string
```
Comment (stdweird) [inline quattor/types/hardware.pan]: validation?

File: quattor/types/hardware.pan [hunk 5]
```diff
@@ -94,6 +124,8 @@ type structure_cards = {
     "sas" ? structure_raid{}
     # For describing BMC controllers
     "bmc" ? structure_bmc[]
+    # For describing IB HCA's
```
Comment (stdweird) [inline quattor/types/hardware.pan]: use annotations (also above)

**Comment by wdpypere** [comment]
@stdweird added documentation and validation. Don't know why panlint fails though. Locally it runs fine, but that might be a different version.

**Comment by jrha** [comment]
Panlint crash should be fixed by https://github.com/quattor/pan/pull/151

File: quattor/functions/validation.pan [hunk 6]
```diff
@@ -21,10 +21,10 @@ function is_component_list = {
     base = "/software/components/";
 
     components = ARGV[0];
-    ok = first(components,k,v);
+    ok = first(components, k, v);
```
Comment (stdweird) [inline quattor/functions/validation.pan]: i know this is not your code, but can you rework this in a for loop?

File: quattor/functions/validation.pan [hunk 7]
```diff
@@ -72,27 +72,27 @@ function is_profile_list = {
     profiles = ARGV[0];
     errors_ni = '';
     errors_na = '';
-    ok = first(profiles,k,v);
+    ok = first(profiles, k, v);
```
Comment (stdweird) [inline quattor/functions/validation.pan]: same remark, try to make this a for loop

Comment (stdweird) [inline quattor/functions/validation.pan]: actuaully, what does this horrible code try to do? can it be replaced by a format?

File: quattor/functions/validation.pan [hunk 8]
```diff
@@ -129,24 +129,23 @@ function is_card_port = {
...
-     true;
+    l = list ("raid", "ide", "sata", "scsi", "sas");
+
+    cardlist = ARGV[0];
+
+    foreach (i; card; l) {
+        if (exists (cardlist[card])) {
+            foreach (controller; data; cardlist[card]) {
+                if (!match (controller, '^_\d+$')) {
+                    error ("Card " + card + " index " + controller +
```
Comment (stdweird) [inline quattor/functions/validation.pan]: my eyes....
single line format pls.

File: quattor/functions/validation.pan [hunk 9]
```diff
@@ -157,3 +156,38 @@ function valid_absolute_file_paths = {
...
+
+Ex.: 80-06-00-48-FE-80-00-00-00-00-00-00-00-00-00-00
+    80:06:00:48:FE:80:00:00:00:00:00:00:00:00:00:00
+
+Both upper and lower-case hex digits are accepted.
+}
+function is_ibhwaddr = {
+    # Check cardinality and type of argument.
+    if (ARGC != 1 || !is_string(ARGV[0]))
+        error("usage: is_ibguid(string)");
```
Comment (stdweird) [inline quattor/functions/validation.pan]: wrng message

File: quattor/functions/validation.pan [hunk 10]
```diff
@@ -157,3 +156,38 @@ function valid_absolute_file_paths = {
...
+
+    return(match(ARGV[0],
```
Comment (stdweird) [inline quattor/functions/validation.pan]: no `return` needed

File: quattor/functions/validation.pan [hunk 11]
```diff
@@ -157,3 +156,38 @@ function valid_absolute_file_paths = {
...
+The Infiniband guid is a series of 16 chars preceded by 0x.
+
+Ex.: 0x0002c9030002fb06
+
+Both upper and lower-case hex digits are accepted.
+}
+function is_ibguid = {
+    # Check cardinality and type of argument.
+    if (ARGC != 1 || !is_string(ARGV[0]))
+        error("usage: is_ibhwaddr(string)");
```
Comment (stdweird) [inline quattor/functions/validation.pan]: wrong message

File: quattor/functions/validation.pan [hunk 12]
```diff
@@ -157,3 +156,38 @@ function valid_absolute_file_paths = {
...
+
+    return(match(ARGV[0],
```
Comment (stdweird) [inline quattor/functions/validation.pan]: no return

File: quattor/functions/validation.pan [hunk 13]
```diff
@@ -157,3 +156,38 @@ function valid_absolute_file_paths = {
...
+The Infiniband hardware address is a series of twenty bytes encoded as hex values
+and separated with a colon or a hyphen.  Within a value you must
+use a consistent separator.
+
+Ex.: 80-06-00-48-FE-80-00-00-00-00-00-00-00-00-00-00
+    80:06:00:48:FE:80:00:00:00:00:00:00:00:00:00:00
+
+Both upper and lower-case hex digits are accepted.
+}
+function is_ibhwaddr = {
```
Comment (stdweird) [inline quattor/functions/validation.pan]: why a function?

File: quattor/functions/validation.pan [hunk 14]
```diff
@@ -157,3 +156,38 @@ function valid_absolute_file_paths = {
...
+};
+
+@documentation{
+The Infiniband guid is a series of 16 chars preceded by 0x.
+
+Ex.: 0x0002c9030002fb06
+
+Both upper and lower-case hex digits are accepted.
+}
+function is_ibguid = {
```
Comment (stdweird) [inline quattor/functions/validation.pan]: why a function?

File: quattor/types/hardware.pan [hunk 15]
```diff
@@ -75,25 +75,59 @@ type structure_bmc = {
...
+    "pci" ? structure_pci
+    "guid" ? string with is_ibguid (SELF)
+};
+
+@documentation{
     Card and/or addon
 }
 type structure_cards = {
-    # Indexed by device name (eth0, venet0...)
+    @{ desc =  Indexed by device name (eth0, venet0...)}
```
Comment (stdweird) [inline quattor/types/hardware.pan]: what is the advantage of the `desc = `? these are not functions and we do not use this anywhere else in attribute annotation

Comment (wdpypere) [inline quattor/types/hardware.pan]: No advantage, but it is allowed. I'll remove it.

File: quattor/functions/validation.pan (hunk 14 above)
Comment (wdpypere) [inline quattor/functions/validation.pan]: I don't understand the question. why not a function?

File: quattor/functions/validation.pan (hunk 6 above)
Comment (wdpypere) [inline quattor/functions/validation.pan]: I'll try.

File: quattor/functions/validation.pan (hunk 14 above)
Comment (stdweird) [inline quattor/functions/validation.pan]: if you only need a type, make a type. in panc 10.4, you'll be able to make test functions from any type if you'd ever need one.

**Comment by wdpypere** [comment]
@stdweird remarks addressed

**Comment by jrha** [comment]
@hpcugentbot test this please

**Pull request merged**